<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>How to use indigopy &mdash; INDIGOpy  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Prepare data from XLSX file" href="prepare_data_from_xlsx.html" />
    <link rel="prev" title="tests package" href="../tests.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> INDIGOpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../modules.html">INDIGOpy</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">How to use <code class="docutils literal notranslate"><span class="pre">indigopy</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Set-up-environment">Set up environment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example:-E.-coli">Example: <em>E. coli</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example:-M.-tuberculosis">Example: <em>M. tuberculosis</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example:-S.-aureus">Example: <em>S. aureus</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example:-A.-baumannii">Example: <em>A. baumannii</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Extension-to-CARAMeL:-using-metabolic-flux-data-as-drug-profiles">Extension to CARAMeL: using metabolic flux data as drug profiles</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="prepare_data_from_xlsx.html">Prepare data from XLSX file</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">INDIGOpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>How to use <code class="docutils literal notranslate"><span class="pre">indigopy</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/indigopy_demo.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="How-to-use-indigopy">
<h1>How to use <code class="docutils literal notranslate"><span class="pre">indigopy</span></code><a class="headerlink" href="#How-to-use-indigopy" title="Permalink to this heading"></a></h1>
<p>Example code for how to use the <code class="docutils literal notranslate"><span class="pre">indigopy</span></code> package. The sample data used in this example notebook is derived from the <a class="reference external" href="https://doi.org/10.1007/978-1-4939-8891-4_13">INDIGO</a>, <a class="reference external" href="https://doi.org/10.1128/mbio.02627-19">INDIGO-MTB</a>, and <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1006677">MAGENTA</a> publications.</p>
<section id="Set-up-environment">
<h2>Set up environment<a class="headerlink" href="#Set-up-environment" title="Permalink to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Import dependencies
import pandas as pd
from scipy.stats import spearmanr
from sklearn.metrics import r2_score, classification_report
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
import seaborn as sns
import matplotlib.pyplot as plt

# Import package functions
from indigopy.core import load_sample, featurize, classify
</pre></div>
</div>
</div>
</section>
<section id="Example:-E.-coli">
<h2>Example: <em>E. coli</em><a class="headerlink" href="#Example:-E.-coli" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The following analysis and results were originally reported in the <a class="reference external" href="https://doi.org/10.1007/978-1-4939-8891-4_13">INDIGO</a> paper.</div>
<div class="line">- <strong>Training dataset</strong>: 105 two-way interactions between 15 antibiotics</div>
<div class="line">- <strong>Testing dataset</strong>: 66 two-way interactions between the 15 antibiotics in the training set + 4 new antibiotics</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load sample data
sample = load_sample(&#39;ecoli&#39;)

# Define input arguments
key             = sample[&#39;key&#39;]
profiles        = sample[&#39;profiles&#39;]
feature_names   = sample[&#39;feature_names&#39;]
train_ixns      = sample[&#39;train&#39;][&#39;interactions&#39;]
train_scores    = sample[&#39;train&#39;][&#39;scores&#39;]
test_ixns       = sample[&#39;test&#39;][&#39;interactions&#39;]
test_scores     = sample[&#39;test&#39;][&#39;scores&#39;]

# Determine ML features
train_data      = featurize(train_ixns, profiles, feature_names=feature_names, key=key, silent=True)
test_data       = featurize(test_ixns, profiles, feature_names=feature_names, key=key, silent=True)
X_train, X_test = train_data[&#39;feature_df&#39;].to_numpy().transpose(), test_data[&#39;feature_df&#39;].to_numpy().transpose()

# Determine class labels
thresh, classes = (-0.5, 2), (&#39;S&#39;, &#39;N&#39;, &#39;A&#39;)
train_labels    = classify(train_scores, thresholds=thresh, classes=classes)
test_labels     = classify(test_scores, thresholds=thresh, classes=classes)

# Train and apply a regression-based model
reg_model = RandomForestRegressor()
reg_model.fit(X_train, train_scores)
reg_y = reg_model.predict(X_test)
r1, p = spearmanr(test_scores, reg_y)
r2 = r2_score(test_scores, reg_y)
print(&#39;Regression results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(r1, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
print(&#39;\tR2 = {}&#39;.format(round(r2, 4)))

# Train and apply a classification-based model
class_model = RandomForestClassifier()
class_model.fit(X_train, train_labels)
class_y = class_model.predict(X_test)
print(&#39;Classification results:&#39;)
print(classification_report(test_labels, class_y))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Defining INDIGO features: 100%|██████████| 105/105 [00:00&lt;00:00, 424.50it/s]
Defining INDIGO features: 100%|██████████| 66/66 [00:00&lt;00:00, 441.80it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Regression results:
        Spearman R = 0.6791
        Spearman p = 3.68e-10
        R2 = 0.4068
Classification results:
              precision    recall  f1-score   support

           A       0.44      0.31      0.36        13
           N       0.68      0.90      0.78        42
           S       1.00      0.09      0.17        11

    accuracy                           0.65        66
   macro avg       0.71      0.43      0.44        66
weighted avg       0.69      0.65      0.59        66

</pre></div></div>
</div>
</section>
<section id="Example:-M.-tuberculosis">
<h2>Example: <em>M. tuberculosis</em><a class="headerlink" href="#Example:-M.-tuberculosis" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The following analysis and results were originally reported in the <a class="reference external" href="https://doi.org/10.1128/mbio.02627-19">INDIGO-MTB</a> paper.</div>
<div class="line">- <strong>Training dataset</strong>: 196 two- to five-way interactions between 40 antibacterials</div>
<div class="line">- <strong>Testing dataset</strong>: 36 two- to three-way interactions between the 13 antibacterials</div>
<div class="line">- <strong>Clinical dataset</strong>: clinical outcomes for 57 two- to five-way interactions between 7 antibacterials</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load sample data
sample = load_sample(&#39;mtb&#39;)

# Define input arguments
key             = sample[&#39;key&#39;]
profiles        = sample[&#39;profiles&#39;]
feature_names   = sample[&#39;feature_names&#39;]
train_ixns      = sample[&#39;train&#39;][&#39;interactions&#39;]
train_scores    = sample[&#39;train&#39;][&#39;scores&#39;]
test_ixns       = sample[&#39;test&#39;][&#39;interactions&#39;]
test_scores     = sample[&#39;test&#39;][&#39;scores&#39;]
clinical_ixns   = sample[&#39;clinical&#39;][&#39;interactions&#39;]
clinical_scores = sample[&#39;clinical&#39;][&#39;scores&#39;]

# Determine ML features
train_data      = featurize(train_ixns, profiles, feature_names=feature_names, key=key, silent=True)
test_data       = featurize(test_ixns, profiles, feature_names=feature_names, key=key, silent=True)
clinical_data   = featurize(clinical_ixns, profiles, feature_names=feature_names, key=key, silent=True)
X_train, X_test = train_data[&#39;feature_df&#39;].to_numpy().transpose(), test_data[&#39;feature_df&#39;].to_numpy().transpose()
X_clinical      = clinical_data[&#39;feature_df&#39;].to_numpy().transpose()

# Determine class labels
thresh, classes = (0.9, 1.1), (&#39;S&#39;, &#39;N&#39;, &#39;A&#39;)
train_labels    = classify(train_scores, thresholds=thresh, classes=classes)
test_labels     = classify(test_scores, thresholds=thresh, classes=classes)

# Train and apply a regression-based model
reg_model = RandomForestRegressor()
reg_model.fit(X_train, train_scores)
reg_y = reg_model.predict(X_test)
r, p = spearmanr(test_scores, reg_y)
r2 = r2_score(test_scores, reg_y)
print(&#39;Regression results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(r, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
print(&#39;\tR2 = {}&#39;.format(round(r2, 4)))

# Train and apply a classification-based model
class_model = RandomForestClassifier()
class_model.fit(X_train, train_labels)
class_y = class_model.predict(X_test)
print(&#39;Classification results:&#39;)
print(classification_report(test_labels, class_y))

# Apply model to clinical data
clinical_y = reg_model.predict(X_clinical)
r, p = spearmanr(clinical_scores, clinical_y)
print(&#39;Clinical results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(-r, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Defining INDIGO features: 100%|██████████| 196/196 [00:00&lt;00:00, 346.44it/s]
Defining INDIGO features: 100%|██████████| 36/36 [00:00&lt;00:00, 513.40it/s]
Defining INDIGO features: 100%|██████████| 57/57 [00:00&lt;00:00, 491.74it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Regression results:
        Spearman R = 0.5264
        Spearman p = 0.000975
        R2 = 0.0905
Classification results:
              precision    recall  f1-score   support

           A       0.67      0.38      0.48        16
           N       0.00      0.00      0.00         1
           S       0.73      0.84      0.78        19

    accuracy                           0.61        36
   macro avg       0.46      0.41      0.42        36
weighted avg       0.68      0.61      0.63        36

Clinical results:
        Spearman R = 0.3955
        Spearman p = 0.00232
</pre></div></div>
</div>
</section>
<section id="Example:-S.-aureus">
<h2>Example: <em>S. aureus</em><a class="headerlink" href="#Example:-S.-aureus" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The following analysis and results were originally reported in the <a class="reference external" href="https://doi.org/10.1007/978-1-4939-8891-4_13">INDIGO</a> paper.</div>
<div class="line">- <strong>Training dataset</strong>: 171 two-way interactions between 19 antibiotics measured in <em>E. coli</em></div>
<div class="line">- <strong>Testing dataset</strong>: 45 two-way interactions between the 10 antibiotics measured in <em>S. aureus</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load sample data
sample = load_sample(&#39;saureus&#39;)

# Define input arguments
key             = sample[&#39;key&#39;]
profiles        = sample[&#39;profiles&#39;]
feature_names   = sample[&#39;feature_names&#39;]
train_ixns      = sample[&#39;train&#39;][&#39;interactions&#39;]
train_scores    = sample[&#39;train&#39;][&#39;scores&#39;]
test_ixns       = sample[&#39;test&#39;][&#39;interactions&#39;]
test_scores     = sample[&#39;test&#39;][&#39;scores&#39;]
strains         = sample[&#39;orthology&#39;][&#39;strains&#39;]
orthology_map   = sample[&#39;orthology&#39;][&#39;map&#39;]

# Determine ML features
train_data      = featurize(train_ixns, profiles, feature_names=feature_names, key=key, silent=True)
test_data       = featurize(test_ixns, profiles, feature_names=feature_names, key=key, silent=True,
                            strains=strains, orthology_map=orthology_map)
X_train, X_test = train_data[&#39;feature_df&#39;].to_numpy().transpose(), test_data[&#39;feature_df&#39;].to_numpy().transpose()

# Determine class labels
thresh, classes = (-0.5, 2), (&#39;S&#39;, &#39;N&#39;, &#39;A&#39;)
train_labels    = classify(train_scores, thresholds=thresh, classes=classes)
test_labels     = classify(test_scores, thresholds=thresh, classes=classes)

# Train and apply a regression-based model
reg_model = RandomForestRegressor()
reg_model.fit(X_train, train_scores)
reg_y = reg_model.predict(X_test)
r, p = spearmanr(test_scores, reg_y)
r2 = r2_score(test_scores, reg_y)
print(&#39;Regression results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(r, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
print(&#39;\tR2 = {}&#39;.format(round(r2, 4)))

# Train and apply a classification-based model
class_model = RandomForestClassifier()
class_model.fit(X_train, train_labels)
class_y = class_model.predict(X_test)
print(&#39;Classification results:&#39;)
print(classification_report(test_labels, class_y))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Defining INDIGO features: 100%|██████████| 171/171 [00:00&lt;00:00, 524.04it/s]
Defining INDIGO features: 100%|██████████| 45/45 [00:00&lt;00:00, 412.21it/s]
Mapping orthologous genes: 100%|██████████| 1/1 [00:02&lt;00:00,  2.55s/it]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Regression results:
        Spearman R = 0.4781
        Spearman p = 0.000894
        R2 = -1.1485
Classification results:
              precision    recall  f1-score   support

           A       0.00      0.00      0.00         2
           N       0.49      1.00      0.66        22
           S       0.00      0.00      0.00        21

    accuracy                           0.49        45
   macro avg       0.16      0.33      0.22        45
weighted avg       0.24      0.49      0.32        45

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\Users\carol\AppData\Local\Programs\PythonCodingPack\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
</pre></div></div>
</div>
</section>
<section id="Example:-A.-baumannii">
<h2>Example: <em>A. baumannii</em><a class="headerlink" href="#Example:-A.-baumannii" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The following analysis and results were originally reported in the <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1006677">MAGENTA</a> paper.</div>
<div class="line">- <strong>Training dataset</strong>: 338 two- to three-way interactions between 24 antibiotics measured in <em>E. coli</em> cultured in various media conditions</div>
<div class="line">- <strong>Testing dataset</strong>: 45 two-way interactions between the 8 antibiotics measured in <em>A. baumannii</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load sample data
sample = load_sample(&#39;abaumannii&#39;)

# Define input arguments
key             = sample[&#39;key&#39;]
profiles        = sample[&#39;profiles&#39;]
feature_names   = sample[&#39;feature_names&#39;]
train_ixns      = sample[&#39;train&#39;][&#39;interactions&#39;]
train_scores    = sample[&#39;train&#39;][&#39;scores&#39;]
test_ixns       = sample[&#39;test&#39;][&#39;interactions&#39;]
test_scores     = sample[&#39;test&#39;][&#39;scores&#39;]
strains         = sample[&#39;orthology&#39;][&#39;strains&#39;]
orthology_map   = sample[&#39;orthology&#39;][&#39;map&#39;]

# Determine ML features
train_data      = featurize(train_ixns, profiles, feature_names=feature_names, key=key, silent=True)
test_data       = featurize(test_ixns, profiles, feature_names=feature_names, key=key, silent=True,
                            strains=strains, orthology_map=orthology_map)
X_train, X_test = train_data[&#39;feature_df&#39;].to_numpy().transpose(), test_data[&#39;feature_df&#39;].to_numpy().transpose()

# Determine class labels
thresh, classes = (-0.5, 0), (&#39;S&#39;, &#39;N&#39;, &#39;A&#39;)
train_labels    = classify(train_scores, thresholds=thresh, classes=classes)
test_labels     = classify(test_scores, thresholds=thresh, classes=classes)

# Train and apply a regression-based model
reg_model = RandomForestRegressor()
reg_model.fit(X_train, train_scores)
reg_y = reg_model.predict(X_test)
r, p = spearmanr(test_scores, reg_y)
r2 = r2_score(test_scores, reg_y)
print(&#39;Regression results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(r, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
print(&#39;\tR2 = {}&#39;.format(round(r2, 4)))

# Train and apply a classification-based model
class_model = RandomForestClassifier()
class_model.fit(X_train, train_labels)
class_y = class_model.predict(X_test)
print(&#39;Classification results:&#39;)
print(classification_report(test_labels, class_y))

# Visualize results
df = pd.DataFrame({&#39;x&#39;: test_labels, &#39;y&#39;: reg_y})
df.replace({&#39;A&#39;: &#39;Antagonism&#39;, &#39;N&#39;: &#39;Neutral&#39;, &#39;S&#39;: &#39;Synergy&#39;}, inplace=True)
sns.set(rc={&#39;figure.figsize&#39;:(3, 4)})
ax = sns.boxplot(x=&#39;x&#39;, y=&#39;y&#39;, data=df, order=[&#39;Antagonism&#39;, &#39;Neutral&#39;, &#39;Synergy&#39;], )
ax.set(title=&#39;A. baumannii&#39;, xlabel=&#39;True class&#39;, ylabel=&#39;Predicted score&#39;)
plt.show()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Defining INDIGO features: 100%|██████████| 338/338 [00:00&lt;00:00, 457.89it/s]
Defining INDIGO features: 100%|██████████| 45/45 [00:00&lt;00:00, 495.31it/s]
Mapping orthologous genes: 100%|██████████| 1/1 [00:02&lt;00:00,  2.90s/it]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Regression results:
        Spearman R = 0.5987
        Spearman p = 1.4e-05
        R2 = -0.4241
Classification results:
              precision    recall  f1-score   support

           A       0.40      0.94      0.56        17
           N       0.00      0.00      0.00        11
           S       0.80      0.24      0.36        17

    accuracy                           0.44        45
   macro avg       0.40      0.39      0.31        45
weighted avg       0.45      0.44      0.35        45

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\Users\carol\AppData\Local\Programs\PythonCodingPack\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_indigopy_demo_9_3.png" src="../_images/notebooks_indigopy_demo_9_3.png" />
</div>
</div>
</section>
<section id="Extension-to-CARAMeL:-using-metabolic-flux-data-as-drug-profiles">
<h2>Extension to CARAMeL: using metabolic flux data as drug profiles<a class="headerlink" href="#Extension-to-CARAMeL:-using-metabolic-flux-data-as-drug-profiles" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The following example shows how to use metabolic flux data to define the drug profiles, which was introduced in our <a class="reference external" href="https://github.com/sriram-lab/CARAMeL">CARAMeL algorithm</a>. Here we re-run the <em>E. coli</em> example using already saved and normalized metabolic flux data (<code class="docutils literal notranslate"><span class="pre">ecoli_flux_normalized.csv</span></code>), which was simulated using the <a class="reference external" href="https://opencobra.github.io/cobratoolbox/stable/">COBRA Toolbox</a> in <a class="reference external" href="https://www.mathworks.com/products/matlab.html">MATLAB</a>. Please refer to the <a class="reference external" href="https://doi.org/10.1093/pnasnexus/pgac132">CARAMeL
publication</a> for further details.</div>
<div class="line">- <strong>Training dataset</strong>: 105 two-way interactions between 15 antibiotics</div>
<div class="line">- <strong>Testing dataset</strong>: 66 two-way interactions between the 15 antibiotics in the training set + 4 new antibiotics</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load sample data
sample = load_sample(&#39;ecoli&#39;)

# Define input arguments
key             = sample[&#39;key&#39;]
train_ixns      = sample[&#39;train&#39;][&#39;interactions&#39;]
train_scores    = sample[&#39;train&#39;][&#39;scores&#39;]
test_ixns       = sample[&#39;test&#39;][&#39;interactions&#39;]
test_scores     = sample[&#39;test&#39;][&#39;scores&#39;]

# Define drug profile information
df              = pd.read_csv(&#39;ecoli_flux_normalized.csv&#39;)
profiles        = df.loc[:, ~df.columns.isin([&#39;rxn&#39;, &#39;rxnName&#39;, &#39;subSystem&#39;])].to_dict(&#39;list&#39;)
feature_names   = df[&#39;rxn&#39;].tolist()

# Determine ML features
train_data      = featurize(train_ixns, profiles, feature_names=feature_names, key=key, silent=True)
test_data       = featurize(test_ixns, profiles, feature_names=feature_names, key=key, silent=True)
X_train, X_test = train_data[&#39;feature_df&#39;].to_numpy().transpose(), test_data[&#39;feature_df&#39;].to_numpy().transpose()

# Determine class labels
thresh, classes = (-0.5, 2), (&#39;S&#39;, &#39;N&#39;, &#39;A&#39;)
train_labels    = classify(train_scores, thresholds=thresh, classes=classes)
test_labels     = classify(test_scores, thresholds=thresh, classes=classes)

# Train and apply a regression-based model
reg_model = RandomForestRegressor()
reg_model.fit(X_train, train_scores)
reg_y = reg_model.predict(X_test)
r, p = spearmanr(test_scores, reg_y)
r2 = r2_score(test_scores, reg_y)
print(&#39;Regression results:&#39;)
print(&#39;\tSpearman R = {}&#39;.format(round(r, 4)))
print(&#39;\tSpearman p = {:.3g}&#39;.format(p))
print(&#39;\tR2 = {}&#39;.format(round(r2, 4)))

# Train and apply a classification-based model
class_model = RandomForestClassifier()
class_model.fit(X_train, train_labels)
class_y = class_model.predict(X_test)
print(&#39;Classification results:&#39;)
print(classification_report(test_labels, class_y))

# Visualize compare INDIGO vs. CARAMeL
df = pd.DataFrame({&#39;Algorithm&#39;: [&#39;INDIGO&#39;, &#39;CARAMeL&#39;], &#39;Spearman R&#39;: [r1, r]})
sns.set(rc={&#39;figure.figsize&#39;:(3, 4)})
ax = sns.barplot(x=&#39;Algorithm&#39;, y=&#39;Spearman R&#39;, data=df, order=[&#39;INDIGO&#39;, &#39;CARAMeL&#39;])
ax.set(title=&#39;E. coli results comparison&#39;)
ax.bar_label(ax.containers[0])
plt.show()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Defining INDIGO features: 100%|██████████| 105/105 [00:00&lt;00:00, 530.35it/s]
Defining INDIGO features: 100%|██████████| 66/66 [00:00&lt;00:00, 515.35it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Regression results:
        Spearman R = 0.7238
        Spearman p = 6.57e-12
        R2 = 0.4797
Classification results:
              precision    recall  f1-score   support

           A       0.53      0.69      0.60        13
           N       0.72      0.79      0.75        42
           S       0.67      0.18      0.29        11

    accuracy                           0.67        66
   macro avg       0.64      0.55      0.55        66
weighted avg       0.67      0.67      0.64        66

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_indigopy_demo_11_2.png" src="../_images/notebooks_indigopy_demo_11_2.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tests.html" class="btn btn-neutral float-left" title="tests package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="prepare_data_from_xlsx.html" class="btn btn-neutral float-right" title="Prepare data from XLSX file" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chandrasekaran Lab, University of Michigan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>